[Merged by Bors] - feat (CategoryTheory): Chosen pullbacks

sinhp opened the pull request:
We develop a computable implementation of pullbacks, by introducing a new type-class `ChosenPullbacksAlong`. The non-computable `HasPullbacksAlong` and `HasPullbacks` predicates yield instances of `ChosenPullbacksAlong` (using global choice), but interestingly in the category of types every morphism has chosen pullbacks. Also, we prove that in cartesian monoidal categories, morphisms to the terminal object and the product projections have chosen pullbacks. We prove that `ChosenPullbacksAlong` has good closure properties, e.g., id, isos have chosen pullbacks, and composition of morphisms with chosen pullbacks have chosen pullbacks.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sinhp committed:
ChosenPullback

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,200 @@
+namespace ChosenPullback
+
+/-- `ChosenPullback (Over.mk f).hom` from  `ChosenPullback f`. -/
+def ofOverMk {Y X : C} (f : Y ⟶ X) [ChosenPullback f] : ChosenPullback (Over.mk f).hom :=
```
```suggestion
def overMkHom {Y X : C} (f : Y ⟶ X) [ChosenPullback f] : ChosenPullback (Over.mk f).hom :=
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,200 @@
+  Over.homMk (snd f g)
+
+theorem snd'_left : (snd' f g).left = snd f g := by
+  rfl
```
```suggestion
theorem snd'_left : (snd' f g).left = snd f g := rfl
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,200 @@
+    simpa using h₁
+  have : φ₁' = φ₂' := by
+    apply (adj.homEquiv U _).symm.injective
+    exact this
```
Is it possible to move the proof part of the first `have` in place of the `exact this` in the second `have`?

joelriou replied:
Also, the `simp` is nonterminal.

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,200 @@
+  have : (Over.map g).map l' ≫ fst' f g = a' := by
+    simp [← Adjunction.homEquiv_counit]
+    aesop
+  apply congr_arg CommaMorphism.left this
```
Untested:
```suggestion
  exact congr_arg CommaMorphism.left this
```

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
pullbackObj

sinhp committed:
pullbackObj

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
ChosenPullback fst

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
ChosenPullback.hasPullbackAlong

sinhp committed:
ChosenPullback.pullback iso Over.pullback

sinhp committed:
remove second have

sinhp committed:
rm simp

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
rm have

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,215 @@
+  Over.w (fst' f g)
+
+variable (f g) in
+
```
```suggestion
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,215 @@
+  let a' : (Over.map g).obj (Over.mk b) ⟶ Over.mk f := Over.homMk a h
+  have : (Over.map g).map (adj.homEquiv (.mk b) (.mk f) (Over.homMk a)) ≫ fst' f g = a' := by
+    simp [← Adjunction.homEquiv_counit]
+    aesop
```
This nonterminal `simp` should be "squeezed". Could you apply `simp?`?

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,215 @@
+/-- The computable `ChosenPullback.pullback g` is naturally isomorphic to the noncomputable
+`Over.pullback g`. -/
+@[simps!]
+noncomputable def pullback_iso_overPullback : ChosenPullback.pullback g ≅ Over.pullback g :=
```
```suggestion
noncomputable def pullbackIsoOverPullback : ChosenPullback.pullback g ≅ Over.pullback g :=
```

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
naming

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
terminal simp

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
rm simp!

sinhp committed:
ChosenPullbacks C

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,228 @@
+/-- `ChosenPullback (Over.mk f).hom` from  `ChosenPullback f`. -/
+def overMkHom {Y X : C} (f : Y ⟶ X) [ChosenPullback f] : ChosenPullback (Over.mk f).hom :=
+  inferInstanceAs (ChosenPullback f)
+
```
I think this can be removed:
```suggestion
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,228 @@
+lemma pullbackIsoOverPullback_hom_app_comp_snd (Z : Over X) :
+    ((pullbackIsoOverPullback g).hom.app Z).left ≫ pullback.snd _ _ = snd _ _ :=
+  Over.w ((pullbackIsoOverPullback g).hom.app Z)
+
```
```suggestion
```

sinhp committed:
HasPullbacks instance from ChosenPullbacks

sinhp committed:
Update ChosenPullback.lean

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
remove overMkHom

sinhp committed:
doc string

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
doc string added

sinhp committed:
Create types.lean

sinhp committed:
rm types from this

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
Create Types.lean

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/Types.lean

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,252 @@
+def pullbackObjMap {Y' : C} (h : Y' ⟶ Y) : pullbackObj (h ≫ f) g ⟶ pullbackObj f g :=
+  lift (fst _ _ ≫ h) (snd _ _) (by simp [condition])
+
+@[reassoc]
```
```suggestion
@[reassoc (attr := simp)]
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,252 @@
+
+section Lift
+
+variable {W : C} (a : W ⟶ Y) (b : W ⟶ Z) (h : a ≫ f = b ≫ g)
```
```suggestion
variable {W : C} (a : W ⟶ Y) (b : W ⟶ Z) (h : a ≫ f = b ≫ g := by cat_disch)
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,252 @@
+  mapPullbackAdj.unit.app Q :=
+    Over.homMk (lift (𝟙 _) (Q.hom ≫ snd _ _) (by simp [condition]))
+    (by simp only [pullbackObjMap]; aesop)
+  mapPullbackAdj.counit.app W := Over.homMk (fst _ g) (by simp [pullbackObjMap])
```
```suggestion
attribute [local simp] condition in
/-- If `g` has a chosen pullback, then `Over.ChosenPullback.fst f g` has a chosen pullback. -/
def chosenPullbackOfFst : ChosenPullback (fst f g) where
  pullback.obj W := Over.mk (pullbackObjMap W.hom)
  pullback.map {W' W} k := Over.homMk (lift (fst _ g ≫ k.left) (snd _ g))
  mapPullbackAdj.unit.app Q := Over.homMk (lift (𝟙 _) (Q.hom ≫ snd _ _))
  mapPullbackAdj.counit.app W := Over.homMk (fst _ g)
```

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
pullbackIsoOverPullback_inv_app_

sinhp commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,249 @@
+  (ChosenPullback.mapPullbackAdj g).rightAdjointUniq (Over.mapPullbackAdj g)
+
+@[reassoc (attr := simp)]
+theorem pullbackIsoOverPullback_hom_app_comp_fst (Z : Over X) :
```
added in the last commit.

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,262 @@
+  simp [lift]
+
+/-- The functoriality of `pullbackObj f g` in the first argument. -/
+def pullbackObjMap {Y' : C} (h : Y' ⟶ Y) : pullbackObj (h ≫ f) g ⟶ pullbackObj f g :=
```
This should probably be generalized to look similar to `pullback.map`.

sinhp committed:
generalize pullbackMap

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
simp nf

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,312 @@
+/-- The canonical morphism `pullbackObj f g ⟶ pullbackObj (f ≫ i) (g ≫ i)`. -/
+def pullbackMapDesc {X' : C} (i : X ⟶ X') [ChosenPullback (g ≫ i)] :
+    pullbackObj f g ⟶ pullbackObj (f ≫ i) (g ≫ i) :=
+  pullbackMap (f ≫ i) (g ≫ i) f g (𝟙 _) (𝟙 _) i
```
I would suggest removing this as it is no longer very much required for the definitions below.

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,312 @@
+    {δ₁ : Y'' ⟶ Y'} {δ₂ : Z'' ⟶ Z'} {δ₃ : X'' ⟶ X'}
+    (comm₁ comm₂ comm₁' comm₂' := by cat_disch) :
+    pullbackMap f' g' f'' g'' δ₁ δ₂ δ₃ comm₁' comm₂' ≫
+    pullbackMap f g f' g'  γ₁ γ₂ γ₃ comm₁ comm₂ =
```
```suggestion
      pullbackMap f g f' g'  γ₁ γ₂ γ₃ comm₁ comm₂ =
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,312 @@
+  w := condition
+  isLimit' :=
+    ⟨PullbackCone.IsLimit.mk _ (fun s ↦ lift s.fst s.snd s.condition)
+      (by simp) (by simp) (by aesop)⟩
```
Untested, but `cat_disch` is preferred over `aesop` in category theory:
```suggestion
      (by simp) (by simp) (by cat_disch)⟩
```

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,312 @@
+@[reassoc (attr := simp)]
+theorem pullbackIsoOverPullback_inv_app_comp_fst (T : Over X) :
+    ((pullbackIsoOverPullback g).inv.app T).left ≫ fst _ _ = pullback.fst _ _ := by
+  convert (Over.forget _).congr_map
```
do you mean using `pullbackIsoOverPullback_hom_app_comp_fst` in this proof or imitating a proof of `pullbackIsoOverPullback_hom_app_comp_fst`? I used `simpa using ...` but somehow it did not work.

sinhp replied:
managed to construct a proof without convert using `pullbackIsoOverPullback_hom_app_comp_fst` in the proof.

sinhp committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into lccc-chosen_pullback

sinhp committed:
rm mapPullbackDesc

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
proof wo convert

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
doc string added

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
use (Over.forget ..).mapIso

sinhp commented on Mathlib/CategoryTheory/Comma/Over/Basic.lean:
```diff
@@ -123,6 +123,16 @@ lemma inv_left_hom_left {f g : Over X} (e : f ≅ g) :
   simp [← Over.comp_left]
 
+/-- An isomorphism of objects in the over category induces an isomorphism of their domains. -/
+def isoLeftIso {f g : Over X} (e : f ≅ g) : f.left ≅ g.left :=
```
done!

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,308 @@
+
+@[simp]
+theorem pullbackMap_id : pullbackMap f g f g (𝟙 Y) (𝟙 Z) (𝟙 X) = 𝟙 _ := by
+  apply hom_ext <;> simp
```
Untested, but as `hom_ext` is an `ext` lemma, it should reasonably work (and we should investigate how to make this work otherwise!).
```suggestion
  cat_disch
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,308 @@
+      pullbackMap f g f' g'  γ₁ γ₂ γ₃ comm₁ comm₂ =
+    pullbackMap f g f'' g'' (δ₁ ≫ γ₁) (δ₂ ≫ γ₂) (δ₃ ≫ γ₃)
+      (by rw [reassoc_of% comm₁', comm₁, assoc]) (by rw [reassoc_of% comm₂', comm₂, assoc]) := by
+  apply hom_ext <;> simp
```
```suggestion
  cat_disch
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,308 @@
+    (Over.forget ..).mapIso <| pullbackIsoOverPullback g |>.app T
+  have : ((pullbackIsoOverPullback g).inv.app T).left = iso.inv := by rfl
+  simp only [pullback_obj_left, this, Iso.inv_comp_eq]
+  aesop
```
```suggestion
  rw [← pullbackIsoOverPullback_hom_app_comp_fst, ← Over.comp_left_assoc,
    Iso.inv_hom_id_app, id_left, id_comp]
```

sinhp committed:
ext instead of hom_ext

sinhp committed:
change fst to snd

this is because, as seen in lccc-section, we want to have
ChosenPullback.snd (toUnit X) (toUnit I) = CartesianMonoidalCategory.snd X I := rfl

sinhp committed:
doc string

sinhp committed:
naming

sinhp committed:
computable cartesian monoidal for slices

sinhp committed:
pullbackId and pullbackComp

sinhp committed:
isoInv

sinhp commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean:
```diff
@@ -0,0 +1,369 @@
+end PullbackFromChosenPullbacks
+
+section
+
```
The reason I included them here is that I need to use a _computable_ cartesian monoidal structure from chosen pullbacks in the PR branch `lccc-basic` to relate having pushforwards to having cartesian closedness in the slices. Without cartesian monoidal from chosen pullbacks one can not even talk about cartesian closedness in the slices. 

What would be a better place than this file for this? I feel the cartesian monoidal structure from chosen pullback is very much related to ChosenPullbacks no?

sinhp replied:
what would be the name of that file? So far the name of the files in the LocallyCartesianClosed folder: 

- `ChosenPullback.lean`
- `Sections.lean`
- `ExponentiableMorphism.lean`
- `Basic.lean`
- `Types.lean`
- `Beck-Chevalley`


(The file connecting the ChosenPullback and the CartesianMonoidalCategory (Over _) should sit between the first and the second in the list above due to dependencies.) 

(my idea was to add them to the _same_ file `ChosenPullback.lean` in a new PR).

sinhp committed:
add docstring

sinhp committed:
Update ChosenPullback.lean

sinhp committed:
cartesian monoidal section moved to a new PR

sinhp committed:
remove namespace Over

sinhp committed:
rename to ChosenPullbackAlong

sinhp committed:
improve docstrings

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

sinhp committed:
add s

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
s

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullbacksAlong.lean:
```diff
@@ -0,0 +1,317 @@
+
+attribute [local simp] condition in
+/-- If `g` has a chosen pullback, then `Over.ChosenPullbacksAlong.fst f g` has a chosen pullback. -/
+def ChosenPullbacksAlongOfFst : ChosenPullbacksAlong (fst f g) where
```
```suggestion
def chosenPullbacksAlongFst : ChosenPullbacksAlong (fst f g) where
```

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullbacksAlong.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullbacksAlong.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullbacksAlong.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullbacksAlong.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
type of

sinhp committed:
simp

joelriou committed:
Apply suggestion from @joelriou
